atlasdb-workload-server/src/test/java/com/palantir/atlasdb/workload/store/UnreliableTimestampManagerTest.java

```diff
@@ -0,0 +1,170 @@
+        try {
+            for (int i = 0; i < 10; i++) {
+                executor.execute(timestampManager::randomlyIncreaseTimestamp);
+            }
+        } finally {
+            executor.shutdown();
+            assertThat(executor.awaitTermination(10, TimeUnit.SECONDS)).isTrue();
+        }
+
+        assertListRandomlyIncreasesInValue(allValues);
+    }
+
+    @ParameterizedTest
+    @MethodSource("timestampMethods")
+    public void gettingTimestampsBlocksOnPendingFastForward(Consumer<TimestampManager> task)
+            throws InterruptedException {
+        CountDownLatch startingFastForward = new CountDownLatch(1);
+        CountDownLatch blockFastForward = new CountDownLatch(1);
+
+        UnreliableTimestampManager timestampManager =
+                UnreliableTimestampManager.create(timelockService, () -> 0L, _value -> {
+                    startingFastForward.countDown();
+                    try {
+                        blockFastForward.await();
+                    } catch (InterruptedException e) {
+                        throw new RuntimeException(e);
+                    }
+                });
+
+        ExecutorService executor = Executors.newCachedThreadPool();
+        try {
+            Future<?> increaseTimestampFuture = executor.submit(timestampManager::randomlyIncreaseTimestamp);
+            startingFastForward.await();
+            Future<?> getTimestamp = executor.submit(() -> task.accept(timestampManager));
+            assertThat(getTimestamp).isNotDone();
+            blockFastForward.countDown();
+            assertThat(increaseTimestampFuture).succeedsWithin(Duration.ofSeconds(1));
+            assertThat(getTimestamp).succeedsWithin(Duration.ofSeconds(1));
+
+        } finally {
+            executor.shutdown();
+            boolean shutdownSuccessfully = executor.awaitTermination(10, TimeUnit.SECONDS);
+            executor.shutdownNow();
+            assertThat(shutdownSuccessfully).isTrue();
+        }
+    }
+
+    private static void assertListRandomlyIncreasesInValue(List<Long> timestamps) {
+        assertThat(timestamps).isSorted();
+        assertThat(new HashSet<>(timestamps)).hasSameSizeAs(timestamps);
+
+        List<Long> differencesBetweenValues = new ArrayList<>();
+        for (int i = 1; i < timestamps.size(); i++) {
+            differencesBetweenValues.add(timestamps.get(i) - timestamps.get(i - 1));
+        }
+
+        // Verifying things increase randomly. We can't assert that every delta is random, since it's possible that at
+        // least one delta is the same as another, but there's a minuscule chance that, in the domain of all longs, we
+        // roll each unique long that we draw, 4 times.
+        assertThat(new HashSet<>(differencesBetweenValues)).hasSizeGreaterThan(differencesBetweenValues.size() / 4);
+    }
+
+    private UnreliableTimestampManager createThreadsafeArrayBackedTimestampManager(List<Long> timestampList) {
```

Comment:
consider: using Collections.synchronizedList (semantically different) and see if it fits the need (with the caveat of iterators being a foot-gun).

Reply:
Discussed offline why Collections.synchronizedList isn't sufficient alone here (and it was a conscious decision not to use it) - there was an assumption when the above comment was written that there'd be a getLast or get(-1) equivalent on the List API, which is unfortunately not true.